### PR TITLE
done

### PR DESCRIPTION
**Pull request by krisma**
everything's working

### CODE REVIEW

File: www/js/app.js [hunk 1]
```diff
@@ -89,29 +89,30 @@ function handleOpenURL(url) {
     
     if (kvList.route == 'join') {
       var INTRO_DONE_KEY = 'intro_done';
+      var HABITICA_REGISTERED_KEY = 'habitica_registered';
+      var REFERRAL_NAVIGATION_KEY = 'referral_navigation';
+      var REFERED_KEY = 'refered';
```
Comment (shankari) [inline www/js/app.js]: nit: refered -> referred

File: www/js/app.js [hunk 2]
```diff
@@ -89,29 +89,30 @@ function handleOpenURL(url) {
...
+      var REFERED_GROUP_ID = 'refered_group_id';
+      var REFERED_USER_ID = 'refered_user_id'
+      window.localStorage.setItem(REFERED_KEY, true);
```
Comment (shankari) [inline www/js/app.js]: why not use the `angularLocalStorage` service? that's what `startprefs.js` does...

File: www/js/app.js [hunk 3]
```diff
@@ -89,29 +89,30 @@ function handleOpenURL(url) {
...
+      window.localStorage.setItem(REFERED_USER_ID, kvList['userid']);
       if (window.localStorage.getItem(INTRO_DONE_KEY) == false) { 
         // User type 1: Completely new user
-        var REFERED_KEY = 'refered';
-        var REFERED_GROUP_ID = 'refered_group_id';
-        window.localStorage.setItem(REFERED_KEY, true);
-        window.localStorage.setItem(REFERED_GROUP_ID, kvList['groupid']);
-        alert(JSON.stringify(kvList));
-      } else if (1) { 
+        window.localStorage.setItem(REFERRAL_NAVIGATION_KEY, 'goals');
```
Comment (shankari) [inline www/js/app.js]: remove this dead code

File: www/js/app.js [hunk 4]
```diff
@@ -89,29 +89,30 @@ function handleOpenURL(url) {
...
+      } else if (window.localStorage.getItem(HABITICA_REGISTERED_KEY) != null) { // INTRO_DONE_KEY == true
         // User type 2: User already has the app, but not yet activated the game
         // maybe redirect to game tab? and alert that you sure you want to join this group?
-        var REFERED_KEY = 'refered';
-        var REFERED_GROUP_ID = 'refered_group_id';
-        window.localStorage.setItem(REFERED_KEY, true);
-        window.localStorage.setItem(REFERED_GROUP_ID, kvList['groupid']);
-        alert(JSON.stringify(kvList));
-      } else {
+        window.localStorage.setItem(REFERRAL_NAVIGATION_KEY, 'goals');
```
Comment (shankari) [inline www/js/app.js]: what is the difference between this and type 3?

File: www/js/goals.js [hunk 5]
```diff
@@ -37,6 +37,66 @@ angular.module('emission.main.goals',['emission.services', 'ngSanitize', 'ngAnim
 		$scope.partyModal = modal;
 	});
 
+	var joinGroupSuccess = function() {
+	   var alertPopup = $ionicPopup.alert({
+	     title: 'Cool!',
+	     template: 'Request has been sent.'
```
Comment (shankari) [inline www/js/goals.js]: Change to "you have successfully joined the group!"

File: www/js/metrics.js [hunk 6]
```diff
@@ -14,6 +14,20 @@ angular.module('emission.main.metrics',['nvd3', 'emission.services', 'ionic-date
     var lastWeekCarbonInt = [];
     var twoWeeksAgoCarbonInt = [];
     var twoWeeksAgoCalories = 0;
+    
+
+    // popup saying u need to register first
+    var handlePendingRefer = function() {
+      storage.removeItem(REFERRAL_NAVIGATION_KEY);
+      var REFERRAL_NAVIGATION_KEY = 'referral_navigation';
```
Comment (shankari) [inline www/js/metrics.js]: why is this defined after it is used?

File: www/js/services.js [hunk 7]
```diff
@@ -58,7 +58,31 @@ angular.module('emission.services', [])
...
+.service('ReferHelper', function($http) {
 
+    this.habiticaRegister = function(groupid, successCallback, errorCallback) {
+        window.cordova.plugins.BEMServerComm.getUserPersonalData("/join.group/"+groupid, successCallback, errorCallback);
+    };
+    this.joinGroup = function(groupid, userid) {
+    
+    // TODO:
+    new Promise(function(resolve, reject) {
+        window.cordova.plugins.BEMServerComm.postUserPersonalData("/join.group/"+groupid, "inviter", userid, resolve, reject);
```
Comment (shankari) [inline www/js/services.js]: the way that promises are supposed to be used is that you pass the promise up until you need to use the result somewhere. That allows for maximum flexibility in scheduling.

File: www/js/app.js (hunk 2 above)
Comment (krisma) [inline www/js/app.js]: it's not in the scale of controller

**Pull request merged**